From 89c65c4dc2132a36e4f0c8ac534cb2454803bf41 Mon Sep 17 00:00:00 2001
From: "Eric S. Raymond" <esr@thyrsus.com>
Date: Thu, 10 May 2012 00:59:30 -0400
Subject: [PATCH] Armor the JSON code against zeroed value or attribute
 pointers.

Should never happen, but having the bailout logic in plavce creates static
invariants that should banish a bunch of Coverity warnings.
---
 json.c |    9 +++++++++
 json.h |    1 +
 2 files changed, 10 insertions(+)

--- a/json.c
+++ b/json.c
@@ -259,6 +259,8 @@ static int json_internal_read_object(con
 	    }
 	    break;
 	case in_attr:
+	    if (pattr == NULL)
+		return JSON_ERR_NULLPTR;
 	    if (*cp == '"') {
 		*pattr++ = '\0';
 		json_debug_trace((1, "Collected attribute name %s\n",
@@ -320,6 +322,8 @@ static int json_internal_read_object(con
 	    }
 	    break;
 	case in_val_string:
+	    if (pval == NULL)
+		return JSON_ERR_NULLPTR;
 	    if (*cp == '\\')
 		state = in_escape;
 	    else if (*cp == '"') {
@@ -334,6 +338,8 @@ static int json_internal_read_object(con
 		*pval++ = *cp;
 	    break;
 	case in_escape:
+	    if (pval == NULL)
+		return JSON_ERR_NULLPTR;
 	    switch (*cp) {
 	    case 'b':
 		*pval++ = '\b';
@@ -364,6 +370,8 @@ static int json_internal_read_object(con
 	    state = in_val_string;
 	    break;
 	case in_val_token:
+	    if (pval == NULL)
+		return JSON_ERR_NULLPTR;
 	    if (isspace(*cp) || *cp == ',' || *cp == '}') {
 		*pval = '\0';
 		json_debug_trace((1, "Collected token value %s.\n", valbuf));
@@ -635,6 +643,7 @@ const /*@observer@*/ char *json_error_st
 	"saw quoted value when expecting nonstring",
 	"didn't see quoted value when expecting string",
 	"other data conversion error",
+	"unexpected null value or attribute pointer",
     };
 
     if (err <= 0 || err >= (int)(sizeof(errors) / sizeof(errors[0])))
--- a/json.h
+++ b/json.h
@@ -102,6 +102,7 @@ void json_enable_debug(int, FILE *);
 #define JSON_ERR_NONQSTRING	19	/* didn't see quoted value when expecting string */
 #define JSON_ERR_MISC		20	/* other data conversion error */
 #define JSON_ERR_BADNUM		21	/* error while parsing a numerical argument */
+#define JSON_ERR_NULLPTR	22	/* unexpected null value or attribute pointer */
 
 /*
  * Use the following macros to declare template initializers for structobject 
